dvc/repo/__init__.py

```diff
@@ -314,48 +314,57 @@ def _collect_graph(self, stages=None):
         G = nx.DiGraph()
         stages = stages or self.stages
         stages = [stage for stage in stages if stage]
-        outs = {}
+        outs = Trie()  # Use trie to efficiently find overlapping outs and deps
 
         for stage in stages:
             for out in stage.outs:
-                if out.path_info in outs:
-                    dup_stages = [stage, outs[out.path_info].stage]
+                out_key = out.path_info.parts
+
+                # Check for dup outs
+                if out_key in outs:
+                    dup_stages = [stage, outs[out_key].stage]
                     raise OutputDuplicationError(str(out), dup_stages)
-                outs[out.path_info] = out
 
-        for stage in stages:
-            for out in stage.outs:
-                for p in out.path_info.parents:
-                    if p in outs:
-                        msg = (
-                            "Paths for outs:\n'{}'('{}')\n'{}'('{}')\n"
-                            "overlap. To avoid unpredictable behaviour, "
-                            "rerun command with non overlapping outs paths."
-                        ).format(
-                            str(outs[p]),
-                            outs[p].stage.relpath,
-                            str(out),
-                            out.stage.relpath,
-                        )
-                        raise OverlappingOutputPathsError(outs[p], out, msg)
+                # Check for overlapping outs
+                if outs.has_subtrie(out_key):
+                    parent = out
+                    overlapping = first(outs.values(prefix=out_key))
```

Comment:
just to double check - does it take into account situations like: `/a/b/test` in  `/a/b/testing`?

Reply:
Yes, it builds on `PathInfo(...).parts`.